File: jimmer-dialect/src/main/java/ydb/jimmer/dialect/transaction/YdbRetryPolicy.java
package ydb.jimmer.dialect.transaction;

import java.sql.SQLException;
import java.util.OptionalLong;
import java.util.Set;

/**
 * Decides whether a YDB error should be retried and how long to wait before the next attempt.
 * Logic is ported one-to-one from {@code spring-ydb-retry}'s {@code YdbRetryPolicy.kt}.
 *
 * <ul>
 *   <li>{@link YdbDelayCalculator#fullJitterMillis} &mdash; {@code Aborted},
 *       {@code Undetermined}</li>
 *   <li>{@link YdbDelayCalculator#equalJitterMillis} on the fast tier &mdash;
 *       {@code Unavailable}, transport errors, {@code CLIENT_GRPC_ERROR}</li>
 *   <li>{@link YdbDelayCalculator#equalJitterMillis} on the slow tier &mdash;
 *       {@code Overloaded}, {@code CLIENT_RESOURCE_EXHAUSTED}</li>
 *   <li>zero delay &mdash; {@code BadSession}, {@code SessionBusy}, {@code SessionExpired}</li>
 * </ul>
 *
 * <p>Not retried here (add handling if your workload needs it): {@code TIMEOUT},
 * {@code CLIENT_DEADLINE_EXPIRED}, {@code PRECONDITION_FAILED}, and other vendor codes.
 * Setting {@code idempotent = true} retries any code returned by this policy, not only
 * {@link #isTransientVendorCode}. {@code SESSION_EXPIRED} is retried with zero backoff only when
 * the call is marked idempotent and is not in the transient set.
 */
public final class YdbRetryPolicy {
    private YdbRetryPolicy() {}

    /**
     * These error codes are retried even when the call is not idempotent.
     */
    public static final Set<Integer> TRANSIENT_VENDOR_CODES = Set.of(
            YdbVendorCode.ABORTED,
            YdbVendorCode.UNAVAILABLE,
            YdbVendorCode.OVERLOADED,
            YdbVendorCode.BAD_SESSION,
            YdbVendorCode.SESSION_BUSY,
            YdbVendorCode.CLIENT_RESOURCE_EXHAUSTED
    );

    /**
     * Returns {@code true} if the vendor code is in the always-retried (transient) set.
     */
    public static boolean isTransientVendorCode(int vendorCode) {
        return TRANSIENT_VENDOR_CODES.contains(vendorCode);
    }

    /**
     * Walks the cause chain of {@code error} looking for a {@link SQLException} with a non-zero
     * vendor code. Mirrors kotlin-exposed's {@code extractVendorCode}.
     *
     * @return non-zero vendor code, or {@code 0} if no YDB-style code was found
     */
    public static int extractVendorCode(Throwable error) {
        Throwable current = error;
        while (current != null) {
            if (current instanceof SQLException sqlException) {
                int vendorCode = sqlException.getErrorCode();
                if (vendorCode != 0) {
                    return vendorCode;
                }
            }
            current = current.getCause();
        }
        return 0;
    }

    public static OptionalLong getNextRetryDelayMs(int vendorCode, int attempt, RetryConfig config) {
        // {@code attempt} is the zero-based index of the attempt that has just failed, so the next
        // attempt is allowed only while we stay within the total {@code maxAttempts} budget
        // (which counts the initial execution).
        if (attempt + 1 >= config.maxAttempts()) {
            return OptionalLong.empty();
        }
        if (vendorCode == 0) {
            return OptionalLong.empty();
        }
        if (!config.idempotent() && !isTransientVendorCode(vendorCode)) {
            return OptionalLong.empty();
        }

        return switch (vendorCode) {
            case YdbVendorCode.BAD_SESSION, YdbVendorCode.SESSION_BUSY, YdbVendorCode.SESSION_EXPIRED ->
                    OptionalLong.of(0L);
            case YdbVendorCode.ABORTED, YdbVendorCode.UNDETERMINED ->
                    OptionalLong.of(YdbDelayCalculator.fullJitterMillis(
                            config.fastBackoffBaseMs(),
                            config.fastCapBackoffMs(),
                            config.fastCeiling(),
                            config.backoffMultiplier(),
                            attempt));
            case YdbVendorCode.UNAVAILABLE, YdbVendorCode.TRANSPORT_UNAVAILABLE, YdbVendorCode.CLIENT_GRPC_ERROR ->
                    OptionalLong.of(YdbDelayCalculator.equalJitterMillis(
                            config.fastBackoffBaseMs(),
                            config.fastCapBackoffMs(),
                            config.fastCeiling(),
                            config.backoffMultiplier(),
                            attempt));
            case YdbVendorCode.OVERLOADED, YdbVendorCode.CLIENT_RESOURCE_EXHAUSTED ->
                    OptionalLong.of(YdbDelayCalculator.equalJitterMillis(
                            config.slowBackoffBaseMs(),
                            config.slowCapBackoffMs(),
                            config.slowCeiling(),
                            config.backoffMultiplier(),
                            attempt));
            default -> OptionalLong.empty();
        };
    }
}

File: kotlin-exposed-dialect/src/main/kotlin/tech/ydb/exposed/dialect/YdbRetryPolicy.kt
package tech.ydb.exposed.dialect

import tech.ydb.exposed.dialect.code.YdbVendorCode
import java.sql.SQLException
import kotlin.math.min
import kotlin.random.Random

/**
 * Retry delay calculation for [ydbTransaction] from JDBC [SQLException.errorCode] (YDB vendor codes).
 *
 * - [fullJitterMillis] — `Aborted`, `Undetermined`
 * - [equalJitterMillis] — `Unavailable`, transport errors, `Overloaded`, resource exhausted
 * - zero delay — `BadSession`, `SessionBusy`, `SessionExpired`
 *
 * Not retried here (add handling if your workload needs it): `TIMEOUT`, `CLIENT_DEADLINE_EXPIRED`,
 * `PRECONDITION_FAILED`, and other vendor codes. [YdbRetryConfig.enableRetryIdempotence] retries
 * any code returned by this policy when `true`, not only [isTransientVendorCode].
 * `SESSION_EXPIRED` is retried with zero backoff but is not in the transient set.
 */
internal fun getNextRetryDelayMs(
    error: Throwable,
    attempt: Int,
    config: YdbRetryConfig,
    random: Random = Random.Default
): Long? {
    if (attempt >= config.maxAttempts - 1) {
        return null
    }

    val vendorCode = extractVendorCode(error) ?: return null

    if (!config.enableRetryIdempotence && !isTransientVendorCode(vendorCode)) {
        return null
    }

    return when (vendorCode) {
        YdbVendorCode.BAD_SESSION,
        YdbVendorCode.SESSION_BUSY,
        YdbVendorCode.SESSION_EXPIRED -> 0L

        YdbVendorCode.ABORTED,
        YdbVendorCode.UNDETERMINED ->
            fullJitterMillis(
                backoffBaseMs = config.fastBackoffBaseMs,
                capMs = config.fastCapBackoffMs,
                ceiling = config.fastCeiling,
                attempt = attempt,
                random = random
            )

        YdbVendorCode.UNAVAILABLE,
        YdbVendorCode.TRANSPORT_UNAVAILABLE,
        YdbVendorCode.CLIENT_GRPC_ERROR ->
            equalJitterMillis(
                backoffBaseMs = config.fastBackoffBaseMs,
                capMs = config.fastCapBackoffMs,
                ceiling = config.fastCeiling,
                attempt = attempt,
                random = random
            )

        YdbVendorCode.OVERLOADED,
        YdbVendorCode.CLIENT_RESOURCE_EXHAUSTED ->
            equalJitterMillis(
                backoffBaseMs = config.slowBackoffBaseMs,
                capMs = config.slowCapBackoffMs,
                ceiling = config.slowCeiling,
                attempt = attempt,
                random = random
            )

        else -> null
    }
}

/** Vendor codes retried without [YdbRetryConfig.enableRetryIdempotence]. */
internal fun isTransientVendorCode(vendorCode: Int): Boolean = vendorCode in TRANSIENT_VENDOR_CODES

internal fun extractVendorCode(error: Throwable): Int? {
    var current: Throwable? = error
    while (current != null) {
        if (current is SQLException) {
            val vendorCode = current.errorCode
            if (vendorCode != 0) {
                return vendorCode
            }
        }
        current = current.cause
    }
    return null
}

internal fun calculateBackoffMillis(
    backoffBaseMs: Int,
    capMs: Int,
    ceiling: Int,
    attempt: Int
): Int = min(backoffBaseMs * (1 shl min(ceiling, attempt)), capMs)

/** Full jitter: uniform in `[0, calculatedBackoff]`. */
internal fun fullJitterMillis(
    backoffBaseMs: Int,
    capMs: Int,
    ceiling: Int,
    attempt: Int,
    random: Random
): Long {
    val calculatedBackoff = calculateBackoffMillis(backoffBaseMs, capMs, ceiling, attempt)
    return random.nextLong(calculatedBackoff + 1L)
}

/** Equal jitter: `calculatedBackoff/2 + calculatedBackoff%2 + random(0..calculatedBackoff/2)`. */
internal fun equalJitterMillis(
    backoffBaseMs: Int,
    capMs: Int,
    ceiling: Int,
    attempt: Int,
    random: Random
): Long {
    val calculatedBackoff = calculateBackoffMillis(backoffBaseMs, capMs, ceiling, attempt)
    val temp = calculatedBackoff / 2
    return temp + calculatedBackoff % 2 + random.nextLong(temp + 1L)
}

private val TRANSIENT_VENDOR_CODES: Set<Int> = setOf(
    YdbVendorCode.ABORTED,
    YdbVendorCode.UNAVAILABLE,
    YdbVendorCode.OVERLOADED,
    YdbVendorCode.CLIENT_RESOURCE_EXHAUSTED,
    YdbVendorCode.BAD_SESSION,
    YdbVendorCode.SESSION_BUSY
)
